propolis-server/src/main.rs

```diff
@@ -0,0 +1,346 @@
+    TypedBody,
+};
+use std::io::{Error, ErrorKind};
+use std::path::PathBuf;
+use std::sync::Arc;
+use structopt::StructOpt;
+
+use propolis::hw::chipset::Chipset;
+use propolis::hw::pci;
+use propolis::instance::Instance;
+
+mod api;
+mod config;
+mod initializer;
+mod serial;
+
+use initializer::{build_instance, MachineInitializer};
+use serial::Serial;
+
+// TODO(error) Do a pass of HTTP codes (error and ok)
+// TODO(idempotency) Idempotency mechanisms?
+
+// All context for a single propolis instance.
+struct InstanceContext {
+    // The instance, which may or may not be instantiated.
+    instance: Arc<Instance>,
+    properties: api::InstanceProperties,
+    serial: Serial,
+}
+
+struct ServerContext {
+    context: tokio::sync::Mutex<Option<InstanceContext>>,
+    config: config::Config,
+}
+
+impl ServerContext {
+    fn new(config: config::Config) -> Self {
+        ServerContext { context: tokio::sync::Mutex::new(None), config }
+    }
+}
+
+fn api_to_propolis_state(
+    state: api::InstanceStateRequested,
+) -> propolis::instance::State {
+    use api::InstanceStateRequested as ApiState;
+    use propolis::instance::State as PropolisState;
+
+    match state {
+        ApiState::Run => PropolisState::Run,
+        ApiState::Stop => PropolisState::Halt,
+        ApiState::Reboot => PropolisState::Reset,
+    }
+}
+
+fn propolis_to_api_state(
+    state: propolis::instance::State,
+) -> api::InstanceState {
+    use api::InstanceState as ApiState;
+    use propolis::instance::State as PropolisState;
+
+    match state {
+        PropolisState::Initialize => ApiState::Creating,
+        PropolisState::Boot => ApiState::Starting,
+        PropolisState::Run => ApiState::Running,
+        PropolisState::Quiesce => ApiState::Stopped,
+        PropolisState::Halt => todo!(),
+        PropolisState::Reset => todo!(),
+        PropolisState::Destroy => ApiState::Destroyed,
+    }
+}
+
+/*
+ * Instances: CRUD API
+ */
+
+#[endpoint {
+    method = PUT,
+    path = "/instances/{instance_id}",
+}]
+async fn instance_ensure(
+    rqctx: Arc<RequestContext<ServerContext>>,
+    path_params: Path<api::InstancePathParams>,
+    request: TypedBody<api::InstanceEnsureRequest>,
+) -> Result<HttpResponseCreated<api::InstanceEnsureResponse>, HttpError> {
+    let server_context = rqctx.context();
+
+    // TODO(idempotency) - how?
+
+    let mut context = server_context.context.lock().await;
+
+    if context.is_some() {
+        return Err(HttpError::for_internal_error(
+            "Server already initialized".to_string(),
+        ));
+    }
+
+    let properties = request.into_inner().properties;
+    if path_params.into_inner().instance_id != properties.id {
+        return Err(HttpError::for_internal_error(
+            "UUID mismatch (path did not match struct)".to_string(),
+        ));
+    }
+
+    // Create the instance.
+    let lowmem = (properties.memory * 1024 * 1024) as usize;
```

Comment:
This is MiB, but the comment in `api::InstanceProeprties` above indicates that field is in MB.

Reply:
yup - updated docs to `MiB`